swarm/storage/mru/request.go

```diff
@@ -117,67 +115,68 @@ func (r *Request) RootAddr() storage.Address {
 
 // StartTime returns the time that the resource was/will be created at
 func (r *Request) StartTime() Timestamp {
-	return r.startTime
+	return r.metadata.StartTime
 }
 
-// OwnerAddr returns the resource owner's address
-func (r *Request) OwnerAddr() common.Address {
-	return r.ownerAddr
+// Owner returns the resource owner's address
+func (r *Request) Owner() common.Address {
+	return r.metadata.Owner
 }
 
 // Sign executes the signature to validate the resource and sets the owner address field
 func (r *Request) Sign(signer Signer) error {
+	if r.metadata.Owner != zeroAddr && r.metadata.Owner != signer.Address() {
+		return NewError(ErrInvalidSignature, "Signer does not match current owner of the resource")
+	}
+
 	if err := r.SignedResourceUpdate.Sign(signer); err != nil {
 		return err
 	}
-	r.ownerAddr = signer.Address()
+	r.metadata.Owner = signer.Address()
 	return nil
 }
 
 // SetData stores the payload data the resource will be updated with
-func (r *Request) SetData(data []byte) {
-	r.signature = nil
+func (r *Request) SetData(data []byte, multihash bool) {
 	r.data = data
-	r.frequency = 0 //mark as update
+	r.multihash = multihash
+	r.dirty()
 }
 
-// decode takes an update request JSON and returns an UpdateRequest
-func (j *updateRequestJSON) decode() (*Request, error) {
-
-	r := &Request{
-		SignedResourceUpdate: SignedResourceUpdate{
-			resourceUpdate: resourceUpdate{
-				updateHeader: updateHeader{
-					UpdateLookup: UpdateLookup{
-						version: j.Version,
-						period:  j.Period,
-					},
-					multihash: j.Multihash,
-				},
-			},
-		},
-		resourceMetadata: resourceMetadata{
-			name:      j.Name,
-			frequency: j.Frequency,
-			startTime: Timestamp{
-				Time: j.StartTime,
-			},
-		},
+func (r *Request) dirty() {
+	r.signature = nil
+	if r.period != 1 || r.version != 1 {
+		r.metadata.Frequency = 0 // mark as update if this is not the first request
 	}
+}
 
-	if err := decodeHexArray(r.startTime.Proof[:], j.StartTimeProof, common.HashLength, "startTimeProof"); err != nil {
-		return nil, err
-	}
+func (r *Request) IsNew() bool {
```

Comment:
Please see further up for comments on this (my review comments got hidden since I intermittently went offline..)